### ISSUE TITLE
[1.x] Cannot validate, "Failed to fetch URL(s) to validate. This may be due to a request timeout."

### ISSUE THREAD

**Issue by jukefr** [ampproject/amp-wp#1371]
What I did :
- 1.0 beta 2 installed
- Paired mode, every content type/template checked, rest is not

When I visit an AMP page via the admin menu or direct link I get directed to `/?amp` then `/?amp-validation-errors-xx` and this redirects me to `/` which means I cannot reach the AMP page.

If I activate the `Automatically accept sanitization for any AMP validation error that is encountered.` option then the `?amp` will work, and not redirect me to `/`.
(I made a separate issue #1369 for this)

In addition to this when I try to see validation results it tells me `Failed to fetch URL(s) to validate. This may be due to a request timeout.` So I can either enable auto-sanitize and not know where the problems are in the first place, but it does work. Or disable auto-sanitize, but I get redirected to `/` and cannot know where my AMP errors are (via `#development=1` or the chrome plugin)

No logs regarding this problem appear in my debug.log so I cannot help much more.

**Comment by westonruter** [comment]
@jukefr The redirection is happening because there are unaccepted validation errors, as you note. So when accessing the `?amp` URL then it redirects to a URL like `?amp_validation_errors=3`, and then this query parameter is stripped from the URL via JavaScript to leave you on the non-AMP page. You should then see in the admin bar:

![image](https://user-images.githubusercontent.com/134745/44742257-439d0800-aab4-11e8-8ac6-eb206969b19c.png)

But you're saying that when you click “Re-validate” it fails with a timeout error. There's nothing in the debug log because the error is right there, a timeout.

What I suggest to try is to increase the timeout for the AMP validation loop requests by putting code like the following in your custom theme's `functions.php` or into a custom plugin:

```php
// Increase timeout for AMP validation requests from 15 to 60 seconds.
add_filter( 'http_request_args', function( $args, $url ) {
	$query = wp_parse_args( wp_parse_url( $url, PHP_URL_QUERY ) );
	if ( isset( $query['amp_validate'] ) ) {
		$args['timeout'] = 60;
	}
	return $args;
}, 10, 2 );
```

Issues related to timeout were previously explored in #1166, where the timeout was increased from 5 seconds to 15 seconds by default. I'm curious how quickly pages are being generated in other cases:

1. How long does it take a page to be generated without AMP?
2. How long does it take a page to be generated with AMP when auto-sanitization is performed? (This is a normal frontend response for visitors.)
3. How long does it take a page to be generated with AMP with validation error sourcing being performed? (This is the request made when updating a post or re-validating a URL for errors.)

The two latter points each involve extra overhead not incurred by the previous two steps. You can manually perform the last step in your browser by making a request with `/?amp&amp_validate` in the URL. You'll be able to to see the JSON blob in an HTML comment at the end of the document for what is then read by the plugin to populate the validation errors in the admin.

**Comment by jukefr** [comment]
Setting a higher timeout works perfectly.
As for your questions (I'll give page local load time instead of actual PHP/WP execution times as it is easier to see):
1. `DOM: 1.81s LOAD: 2.79s FINISH: 3.84s`
2. `DOM:  295ms LOAD: 516ms FINISH: 5.53s`
3. `DOM:  1.09s LOAD: 1.18s FINISH: 1.19s` (being redirected to wp-admin)

**Comment by westonruter** [comment]
The actual PHP/WP execution time is what I need. That is what will cause the timeout.

**Comment by westonruter** [comment]
I'm going to milestone this for 1.0.1 since we may need to increase the timeout generally.

**Comment by westonruter** [comment]
I believe may be largely resolved in 1.5. So I'll close this for now.
